### ISSUE TITLE
Signal.trap("INT") raises error in Ruby 2.5.0-preview1

### ISSUE THREAD

**Issue by danascheider** [bundler/bundler#6192]
Hi there! I just added Ruby 2.5.0-preview1 to my Travis build matrix and started getting build errors when running `bundle --version`. The output of the error is this:
```
$ bundle --version
Traceback (most recent call last):
	3: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/ruby_executable_hooks:15:in `<main>'
	2: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/ruby_executable_hooks:15:in `eval'
	1: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/bundle:23:in `<main>'
/home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/bundle:23:in `load': /home/travis/.rvm/rubies/ruby-2.5.0-preview1/lib/ruby/gems/2.5.0/gems/bundler-1.15.4/exe/bundle:4: syntax error, unexpected tSTRING_BEG, expecting keyword_do or '{' or '(' (SyntaxError)
exec "$bindir/ruby" -x "$0" "$@"
                       ^
/home/travis/.rvm/rubies/ruby-2.5.0-preview1/lib/ruby/gems/2.5.0/gems/bundler-1.15.4/exe/bundle:9: syntax error, unexpected keyword_do_block, expecting end-of-input
Signal.trap("INT") do
                   ^~
The command "bundle --version" failed and exited with 1 during .
Your build has been stopped.
```

I know Ruby 2.5.x is new and don't expect Bundler to support it yet, but figured it might be useful for you to know what errors are being raised. It seems to be raised regardless of other factors in the Travis environment.

I'm definitely interested in helping out with a PR on this issue, but whether I'll be able to will depend on how big a change needs to be made and how much time it takes to identify the problem.

**Comment by indirect** [comment]
Hi there! I'm having some trouble reproducing this. I'm able to run `bundle -v`, and `bundle --version`, and a standalone script that traps INT, all on Ruby 2.5.0 preview 1. Here's what it looks like:

```
$ ruby -v
ruby 2.5.0preview1 (2017-10-10 trunk 60153) [x86_64-darwin17]

$ bundle -v
Bundler version 1.16.0

$ bundle --version
Bundler version 1.16.0

$ cat signal.rb
Signal.trap("INT") do
  puts "interrupted"
  exit(1)
end
loop { sleep 1 }

$ ruby signal.rb
^Cinterrupted
```

I notice in your traceback that you are running bundler via the `ruby_executable_hooks` gem. Could you try uninstalling that gem and running bundler directly? Do you have any other suggestions about how we could reproduce this?

**Comment by hsbt** [comment]
I got a same report on ruby core tracker: https://bugs.ruby-lang.org/issues/13997

It was same as RVM environment.

**Comment by indirect** [comment]
@hsbt thank you for the link, I believe this is the same issue. Did you find why `bundle` has `#!/bin/sh`, but `rake` does not? Is it a default gem difference?

**Comment by hsbt** [comment]
`rake` is not a default gem. Can you try to `rdoc` command with `executable_hooks` ?

**Comment by danascheider** [comment]
@indirect I'm looking into our Travis environment more. I'm not sure why we're using `ruby_executable_hooks` but I'll see if we can uninstall that. I will say that I also haven't managed to repro the issue in my local environment. I'll let you know what I find out. The output that @hsbt linked to is definitely identical to what we're seeing.

**Comment by danascheider** [comment]
Also, I don't know if this is helpful, but here's the totality of the Travis output, so you can see what environment variables are being set and what versions of other things we're using.
```
Worker information
hostname: ip-172-31-6-68:4fdcd8c5-c2b5-48f0-8179-9bb7a8887563
version: v2.7.0 https://github.com/travis-ci/worker/tree/66fc24311e055643710f4e0567f6888c927d3ab8
instance: cb643ee:travis:ruby
startup: 766.902178ms
system_info
Build system information
Build language: ruby
Build group: stable
Build dist: precise
Build id: 755655
Job id: 755763
Build image provisioning date and time
Thu Feb  5 15:09:33 UTC 2015
Operating System Details
Distributor ID:	Ubuntu
Description:	Ubuntu 12.04.5 LTS
Release:	12.04
Codename:	precise
Linux Version
3.13.0-29-generic
Cookbooks Version
a68419e https://github.com/travis-ci/travis-cookbooks/tree/a68419e
GCC version
gcc (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3
Copyright (C) 2011 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
LLVM version
clang version 3.4 (tags/RELEASE_34/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
Pre-installed Ruby versions
ruby-1.9.3-p551
Pre-installed Node.js versions
v0.10.36
Pre-installed Go versions
1.4.1
Redis version
redis-server 2.8.19
riak version
2.0.2
MongoDB version
MongoDB 2.4.12
CouchDB version
couchdb 1.6.1
Neo4j version
1.9.4
RabbitMQ Version
3.4.3
ElasticSearch version
1.4.0
Installed Sphinx versions
2.0.10
2.1.9
2.2.6
Default Sphinx version
2.2.6
Installed Firefox version
firefox 31.0esr
PhantomJS version
1.9.8
ant -version
Apache Ant(TM) version 1.8.2 compiled on December 3 2011
mvn -version
Apache Maven 3.2.5 (12a6b3acb947671f09b81f49094c53f426d8cea1; 2014-12-14T17:29:23+00:00)
Maven home: /usr/local/maven
Java version: 1.7.0_76, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-7-oracle/jre
Default locale: en_US, platform encoding: ANSI_X3.4-1968
OS name: "linux", version: "3.13.0-29-generic", arch: "amd64", family: "unix"
sed: cannot rename /etc/hosts: Device or resource busy
Installing an SSH key from: default repository key
Key fingerprint: 9f:2a:2e:3f:b5:f8:57:1a:4f:c7:d2:a8:2a:89:45:64
git.checkout
2.97s$ git clone --depth=50 git@source.datanerd.us:ruby-agent/ruby_agent.git ruby-agent/ruby_agent
Cloning into 'ruby-agent/ruby_agent'...
Warning: Permanently added 'source.datanerd.us,172.16.128.67' (ECDSA) to the list of known hosts.
remote: Counting objects: 4105, done.
remote: Compressing objects: 100% (1891/1891), done.
remote: Total 4105 (delta 2391), reused 3194 (delta 2127), pack-reused 0
Receiving objects: 100% (4105/4105), 1.21 MiB | 0 bytes/s, done.
Resolving deltas: 100% (2391/2391), done.
Checking connectivity... done.
$ cd ruby-agent/ruby_agent
0.83s$ git fetch origin +refs/pull/1618/merge:
remote: Counting objects: 4, done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 4 (delta 2), reused 3 (delta 2), pack-reused 0
Unpacking objects: 100% (4/4), done.
From source.datanerd.us:ruby-agent/ruby_agent
 * branch            refs/pull/1618/merge -> FETCH_HEAD
$ git checkout -qf FETCH_HEAD
services
0.16s$ sudo service mysql start
start: Job is already running: mysql
0.06s$ sudo service redis-server start
redis-server start/running, process 704
0.02s$ sudo service memcached start
Starting memcached: memcached.
2.41s$ sudo service rabbitmq-server start
 * Starting message broker rabbitmq-server       
Setting environment variables from .travis.yml
$ export BUNDLE_PATH=/home/travis/bundle
$ export MONGODB=2.6.11
$ export JAVA_OPTS="-XX:+TieredCompilation -XX:TieredStopAtLevel=1 -Xmx640m"
$ export JRUBY_OPTS="-Xcompile.invokedynamic=false -J-Djruby.compile.mode=OFF"
$ export RBXOPT="-Xcompiler.no_rbc -Xint"
$ export SERIALIZE=1
$ export TESTOPTS="-v"
$ export TYPE=UNIT
$ export ENVIRONMENT=rails21
rvm
33.29s$ rvm use 2.5.0-preview1 --install --binary --fuzzy
ruby-2.5.0-preview1 is not installed - installing.
Searching for binary rubies, this might take some time.
Found remote file https://s3.amazonaws.com/travis-rubies/binaries/ubuntu/12.04/x86_64/ruby-2.5.0-preview1.tar.bz2
Checking requirements for ubuntu.
Requirements installation successful.
ruby-2.5.0-preview1 - #configure
ruby-2.5.0-preview1 - #download
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 16.1M  100 16.1M    0     0  4146k      0  0:00:03  0:00:03 --:--:-- 4305k
No checksum for downloaded archive, recording checksum in user configuration.
ruby-2.5.0-preview1 - #validate archive
ruby-2.5.0-preview1 - #extract
ruby-2.5.0-preview1 - #validate binary
ruby-2.5.0-preview1 - #setup
ruby-2.5.0-preview1 - #gemset created /home/travis/.rvm/gems/ruby-2.5.0-preview1@global
ruby-2.5.0-preview1 - #importing gemset /home/travis/.rvm/gemsets/global.gems..........................................
ruby-2.5.0-preview1 - #generating global wrappers........
ruby-2.5.0-preview1 - #uninstalling gem rubygems-bundler-1.4.4.
ruby-2.5.0-preview1 - #gemset created /home/travis/.rvm/gems/ruby-2.5.0-preview1
ruby-2.5.0-preview1 - #importing gemset /home/travis/.rvm/gemsets/default.gems....................
ruby-2.5.0-preview1 - #generating default wrappers........
chown: changing ownership of `/home/travis/.rvm/user/installs': Operation not permitted
Using /home/travis/.rvm/gems/ruby-2.5.0-preview1
$ export BUNDLE_GEMFILE=$PWD/Gemfile
$ ruby --version
ruby 2.5.0preview1 (2017-10-10 trunk 60153) [x86_64-linux]
$ rvm --version
rvm 1.26.10 (latest-minor) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]
$ bundle --version
Traceback (most recent call last):
	3: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/ruby_executable_hooks:15:in `<main>'
	2: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/ruby_executable_hooks:15:in `eval'
	1: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/bundle:23:in `<main>'
/home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/bundle:23:in `load': /home/travis/.rvm/rubies/ruby-2.5.0-preview1/lib/ruby/gems/2.5.0/gems/bundler-1.15.4/exe/bundle:4: syntax error, unexpected tSTRING_BEG, expecting keyword_do or '{' or '(' (SyntaxError)
exec "$bindir/ruby" -x "$0" "$@"
                       ^
/home/travis/.rvm/rubies/ruby-2.5.0-preview1/lib/ruby/gems/2.5.0/gems/bundler-1.15.4/exe/bundle:9: syntax error, unexpected keyword_do_block, expecting end-of-input
Signal.trap("INT") do
                   ^~
$ gem --version
2.6.14
before_install.1
0.14s$ gem --version
2.6.14
before_install.2
0.00s$ ./test/script/before_install/gemstash_mirror.sh
if [ -n "$GEMSTASH_MIRROR" ]; then
  bundle config mirror.https://rubygems.org $GEMSTASH_MIRROR
fi
0.11s$ bundle --version
Traceback (most recent call last):
	3: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/ruby_executable_hooks:15:in `<main>'
	2: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/ruby_executable_hooks:15:in `eval'
	1: from /home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/bundle:23:in `<main>'
/home/travis/.rvm/gems/ruby-2.5.0-preview1/bin/bundle:23:in `load': /home/travis/.rvm/rubies/ruby-2.5.0-preview1/lib/ruby/gems/2.5.0/gems/bundler-1.15.4/exe/bundle:4: syntax error, unexpected tSTRING_BEG, expecting keyword_do or '{' or '(' (SyntaxError)
exec "$bindir/ruby" -x "$0" "$@"
                       ^
/home/travis/.rvm/rubies/ruby-2.5.0-preview1/lib/ruby/gems/2.5.0/gems/bundler-1.15.4/exe/bundle:9: syntax error, unexpected keyword_do_block, expecting end-of-input
Signal.trap("INT") do
                   ^~
The command "bundle --version" failed and exited with 1 during .
Your build has been stopped.
```

**Comment by hsbt** [comment]
Can anyone reproduce this without rvm and executable-hooks? I think it's problem of executable-hooks gem.

**Comment by hsbt** [comment]
@danascheider I investigated this without executable-hooks. I found that this is an issue of 2.5.0-preview1, not a bundler.

see. https://bugs.ruby-lang.org/issues/13997#note-10

It will be fixed at 2.5.0-rc1 soon. Please wait a few weeks.

**Comment by danascheider** [comment]
Thanks @hsbt, I appreciate your help!
